File: Stack-Queue-LinkedList/SortStack/SortStack2.java
// only use one additional stack
import java.util.*;
public class SortStack2 {
    // This function return the sorted stack
    public static Stack<Integer> sortstack(Stack<Integer> input) {
        Stack<Integer> tmpStack = new Stack<Integer>();
        while(!input.isEmpty()) {
            int globalMin = Integer.MAX_VALUE;
            while(!input.isEmpty()) {
                int tmpValue = input.pop();
                globalMin = Math.min(globalMin, tmpValue);
                tmpStack.push(tmpValue);
            }
            int count = 0;
            while(!tmpStack.isEmpty() && tmpStack.peek()>=globalMin) {
                int tmpValue = tmpStack.pop();
                if(tmpValue == globalMin) {
                    count++;
                }else {
                    input.push(tmpValue);
                }
            }
            while(count>0) {
                tmpStack.push(globalMin);
                count--;
            }
        }
        return tmpStack;
    }
    
  public static void main(String []args){
    Stack<Integer> input = new Stack<Integer>();
        input.add(34);
        input.add(3);
        input.add(31);
        input.add(98);
        input.add(92);
        input.add(23);
     
        // This is the temporary stack
        Stack<Integer> tmpStack=sortstack(input);
        System.out.println("Sorted numbers are1:");
     
        while (!tmpStack.empty())
        {
            System.out.print(tmpStack.pop()+" ");
        } 
}
}

File: Stack-Queue-LinkedList/SortStack/SortStack2.py
# sort stack using only one additional stack
import sys

class Solution:
  def __init__(self):
    self.stack1 = []

  def sort(self, stack):
    init_length = 0

    while len(stack):
      count = 0
      globla_min = sys.maxsize
      while len(stack):
        tmp = stack.pop()
        globla_min = min(globla_min, tmp)
        self.stack1.append(tmp)
      while len(self.stack1) > init_length:
        tmp = self.stack1.pop()
        if tmp == globla_min:
          count += 1
        else:
          stack.append(tmp)
      while count:
        self.stack1.append(globla_min)
        count -= 1
        
      init_length = len(self.stack1)

    return self.stack1

instance = Solution()
stack = [3,4,5,2,1,2,4,5,6,7,8,8,1]
print(stack)
sorted_stack = instance.sort(stack)
print(sorted_stack)